containers/unit_tests/TestUnorderedSet.hpp

```diff
@@ -0,0 +1,122 @@
+#ifndef KOKKOS_TEST_UNORDERED_SET_HPP
+#define KOKKOS_TEST_UNORDERED_SET_HPP
+
+/**
+ * @file
+ *
+ * The file contains tests that are specific to the specialization of
+ * @ref UnorderedMap as a set.
+ *
+ * Therefore, the number of tests is much smaller than in @ref TestUnorderedMap.hpp, since
+ * only a subset of capabilities are semantically different.
+ *
+ * @note The main difference is that the values view has a zero size for a set.
+ */
+
+#include <random>
+
+#include <gtest/gtest.h>
+
+#include <Kokkos_Random.hpp>
+#include <Kokkos_StdAlgorithms.hpp>
+#include <Kokkos_Sort.hpp>
+#include <Kokkos_UnorderedMap.hpp>
+
+namespace Test {
+
+//! @test Ensure that inserting, erasing and rehasing an unordered set works.
+TEST(TEST_CATEGORY, UnorderedSet_insert_erase_and_rehash) {
+  /**
+   * The set will be inserted for @ref size_all entries,
+   * and then @ref size_erased will be erased from it before rehash.
+   * Once rehasing is done, the capacity of the set is expected to
+   * be smaller than the one of the initial set.
+   */
+  constexpr size_t size_all    = 2579;
+  constexpr size_t size_erased = 568;
+  static_assert(size_all > size_erased);
+
+  using set_type = Kokkos::UnorderedMap<size_t, void, TEST_EXECSPACE>;
+  static_assert(set_type::is_set);
+
+  using policy_type = Kokkos::RangePolicy<TEST_EXECSPACE>;
+  using index_type  = typename policy_type::index_type;
+
+  using index_view_type = Kokkos::View<index_type[size_erased], TEST_EXECSPACE>;
+
+  //! Ensure that the space instance is not the default one.
+  const auto space = Kokkos::Experimental::partition_space(TEST_EXECSPACE{}, 1)[0];
```

Comment:
`partition_space` doesn't guarantee that. 🙂 Also, why does it matter to use an execution space instance different from the default one here?

Reply:
I think at least for `Cuda` and `HIP`, `partition_space` guarantees that we are now on a new stream, that is different from the default one. Am I right?

As for why I'm doing that, it is because in our own code, we noticed that it was the only easy trick to detect code paths executing in a given `space` calling code that runs on the default space, thereby triggering (only from time to time, only on GPUs) a failure in our tests. We already spotted several missing fences in our code thanks to this method.

Reply:
Yes, that's what it does for `Cuda`, `HIP` and `SYCL`.